metricflow-semantics/tests_metricflow_semantics/time/test_time_adjuster.py

```diff
@@ -0,0 +1,156 @@
+    return {
+        TimeGranularity.DAY: 366,
+        TimeGranularity.WEEK: 53,
+        TimeGranularity.MONTH: 31,
+        TimeGranularity.QUARTER: 4,
+        TimeGranularity.YEAR: 1,
+    }
+
+
+def test_start_and_end_periods(  # noqa: D103
+    request: FixtureRequest,
+    mf_test_configuration: MetricFlowTestConfiguration,
+    date_times_to_check: Sequence[datetime.datetime],
+) -> None:
+    pandas_adjuster = PandasTimePeriodAdjuster()
+    dateutil_adjuster = DateutilTimePeriodAdjuster()
+
+    rows: List[Tuple[str, ...]] = []
+    for date_time in date_times_to_check:
+        for time_granularity in TimeGranularity:
+            # Pandas implementation of `adjust_to_start_of_period` doesn't support DAY.
+            if time_granularity == TimeGranularity.DAY:
+                pandas_start_of_period = None
+                pandas_end_of_period = None
+            else:
+                pandas_start_of_period = pandas_adjuster.adjust_to_start_of_period(time_granularity, date_time)
+                pandas_end_of_period = pandas_adjuster.adjust_to_end_of_period(time_granularity, date_time)
+            dateutil_start_of_period = dateutil_adjuster.adjust_to_start_of_period(time_granularity, date_time)
+            dateutil_end_of_period = dateutil_adjuster.adjust_to_end_of_period(time_granularity, date_time)
+            assert (
+                pandas_start_of_period or dateutil_start_of_period
+            ) == dateutil_start_of_period, f"start-of-period mismatch: {date_time.isoformat()} {time_granularity}"
+            assert (
+                pandas_end_of_period or dateutil_end_of_period
+            ) == dateutil_end_of_period, f"end-of-period mismatch: {date_time.isoformat()} {time_granularity}"
+            rows.append(
+                (
+                    date_time.isoformat(),
+                    time_granularity.name,
+                    dateutil_start_of_period.isoformat(),
+                    dateutil_end_of_period.isoformat(),
+                )
+            )
+    assert_str_snapshot_equal(
+        request=request,
+        mf_test_configuration=mf_test_configuration,
+        snapshot_id="results",
+        snapshot_str=tabulate.tabulate(rows, headers=["Date", "Grain", "Period Start", "Period End"]),
+    )
+
+
+def test_expand_time_constraint_to_fill_granularity(  # noqa: D103
+    date_times_to_check: Sequence[datetime.datetime], grain_to_count_in_year: Dict[TimeGranularity, int]
+) -> None:
+    pandas_adjuster = PandasTimePeriodAdjuster()
+    dateutil_adjuster = DateutilTimePeriodAdjuster()
+
+    test_cases = tuple(
+        (start_time, end_time, time_granularity)
+        for start_time in date_times_to_check
+        for time_granularity in TimeGranularity
+        for end_time in (
+            start_time + datetime.timedelta(days=day_offset)
+            for day_offset in range(grain_to_count_in_year[time_granularity] + 2)
```

Comment:
`+2`? Why plus two? To guarantee a spillover across the year boundary?

Reply:
Yeah, added a comment about it.